### DISCUSSION TITLE
Lowest latency settings for using IDEs on high resolution monitors (with both low-ping and high ping connections)?

### DISCUSSION THREAD

**Discussion by jzyrobert** [Xpra-org/xpra#3851] [Q&A] [answered]
Hi,
I have several linux servers in different geographical locations I connect to with Xpra (using SSH) and run Jetbrains IDEs on.
According to the Xpra statistics page, the network latency to one server is around 30ms, and to another is around 100ms.
The bandwidth for these connections should be high enough.

I am passing these flags to the client at the moment:
`--speaker=no --printing=no --notifications=no --compress=0 --speed=100 --opengl=0 --bandwidth-detection=no` and scaling the window to 1440x3440 which gives around 75ms and 160ms of frame total latency for the 30/100ms network latency servers.

Are there additional settings I can tune to reduce this latency further (possibly at cost of graphical fidelity), or will using hardware decoding be required?
Thank you

**Comment by totaam** [comment, marked as answer]
> The bandwidth for these connections should be high enough.

1440x3440 at 30 fps is 3.4Gbps.
How high is it?

> `--compress=0`

Not helpful, `lz4` runs at ram speed.

> `--opengl=0`

Don't do that. This can slow things down dramatically.
Try the opposite: `opengl=force`.

> window to 1440x3440 which gives around 75ms and 160ms of frame total latency for the 30/100ms network latency servers

Those numbers are not bad. That's only ~50ms per frame.

> Are there additional settings I can tune to reduce this latency further (possibly at cost of graphical fidelity)

If you are willing to sacrifice picture quality, use more aggressive `min-quality` and `video-scaling` values.

The best option for low-latency is to use an nvenc capable GPU for video encode. This can halve the frame latency.
Hardware decoding is not usually much of an improvement because decoding is comparatively much easier and even CPUs are fast enough for that.
Perhaps with resolutions closer to 4K, you may see an improvement - there will be some hardware decoding in xpra v5.

**Reply by jzyrobert** [reply]
Thanks, given that the screen in an IDE doesn't change all the time the bandwidth usage I see is nowhere near 3.4Gbps so should be good.
I tested `min-quality=30`, `min-speed=80` and `video-scaling=on` and it seems to provide slightly lower latency numbers.

While on the same topic, what is the reason that websocket based protocols (ws and QUIC) don't provide total frame latency numbers? Is that because of the way they work it can't be calculated?

**Reply by totaam** [reply]
The missing frame latency total was due to this (fixed) bug: 70e3308f4877bc98fe589a4f549e77a07b586f9e
